fe/fe-core/src/main/java/com/starrocks/scheduler/MvTaskRunProcessor.java

```diff
@@ -0,0 +1,500 @@
+        } catch (Exception e) {
+            throw new SemanticException("Expression add partition failed: {}, db: {}, table: {}", e.getMessage(),
+                    database.getFullName(), materializedView.getName());
+        }
+    }
+
+    private void dropPartition(Database database, MaterializedView materializedView, long mvPartitionId) {
+        String dropPartitionName = materializedView.getPartition(mvPartitionId).getName();
+        database.writeLock();
+        try {
+            GlobalStateMgr.getCurrentState().dropPartition(
+                    database, materializedView,
+                    new DropPartitionClause(false, dropPartitionName, false, true));
+            dropPartitionIdRef(materializedView.getPartitionIdRefMap(), mvPartitionId);
+        } catch (Exception e) {
+            throw new SemanticException("Expression drop partition failed: {}, db: {}, table: {}", e.getMessage(),
+                    database.getFullName(), materializedView.getName());
+        } finally {
+            database.writeUnlock();
+        }
+
+    }
+
+    private void addPartitionIdRef(Map<Long, Set<Long>> partitionIdRefMap, long basePartitionId, long mvPartitionId) {
+        partitionIdRefMap.computeIfAbsent(basePartitionId, k -> Sets.newHashSet()).add(mvPartitionId);
+        partitionIdRefMap.computeIfAbsent(mvPartitionId, k -> Sets.newHashSet()).add(basePartitionId);
+    }
+
+    private void dropPartitionIdRef(Map<Long, Set<Long>> partitionIdRefMap, long mvPartitionId) {
+        Set<Long> basePartitionIds = partitionIdRefMap.get(mvPartitionId);
+        for (Long basePartitionId : basePartitionIds) {
+            partitionIdRefMap.get(basePartitionId).remove(mvPartitionId);
+        }
+        partitionIdRefMap.remove(mvPartitionId);
+    }
+
+    private void auditAfterExec(TaskRunContext context, StatementBase parsedStmt, PQueryStatistics statistics) {
+        String origStmt = context.getDefinition();
+        ConnectContext ctx = context.getCtx();
+
+        // slow query
+        long endTime = System.currentTimeMillis();
+        long elapseMs = endTime - ctx.getStartTime();
+
+        ctx.getAuditEventBuilder().setEventType(AuditEvent.EventType.AFTER_QUERY)
+                .setState(ctx.getState().toString()).setErrorCode(ctx.getErrorCode()).setQueryTime(elapseMs)
+                .setScanBytes(statistics == null ? 0 : statistics.scanBytes)
+                .setScanRows(statistics == null ? 0 : statistics.scanRows)
+                .setCpuCostNs(statistics == null || statistics.cpuCostNs == null ? 0 : statistics.cpuCostNs)
+                .setMemCostBytes(statistics == null || statistics.memCostBytes == null ? 0 : statistics.memCostBytes)
+                .setReturnRows(ctx.getReturnRows())
+                .setStmtId(ctx.getStmtId())
+                .setQueryId(ctx.getQueryId() == null ? "NaN" : ctx.getQueryId().toString());
+
+        if (ctx.getState().isQuery()) {
```

Comment:
how do we distinguish the mv queries from the normal queries?

Reply:
You mean we need to do special processing for mv queries?